src/fides/api/models/asset.py

```diff
@@ -0,0 +1,156 @@
+    parent_domain = Column(String)
+    locations = Column(ARRAY(String), server_default="{}", nullable=False)
+    with_consent = Column(BOOLEAN, default=False, nullable=False)
+    data_uses = Column(ARRAY(String), server_default="{}", nullable=False)
+
+    # generic object to store additional attributes, specific to asset type
+    meta = Column(
+        MutableDict.as_mutable(JSONB),
+        nullable=False,
+        server_default="{}",
+        default=dict,
+    )
+
+    # Browser request-specific attributes
+    base_url = Column(String)
+
+    system_id = Column(
+        String, ForeignKey(System.id_field_path, ondelete="CASCADE"), index=True
+    )  # If system is deleted, remove the associated assets.
+
+    system = relationship(
+        System,
+        back_populates="assets",
+        cascade="all,delete",
+        uselist=False,
+        lazy="selectin",
+    )
+
+    # we need to use an md5 of the base_url to avoid constraint/index length issues
+    # and we need to use a unique index, rather than constraint, since postgresql constraints
+    # do not support expressions, only direct column references
+    __table_args__ = (
+        Index(
+            "ix_asset_name_asset_type_domain_base_url_system_id",
+            name,
+            asset_type,
+            domain,
+            func.coalesce(func.md5(base_url), "NULL"),
+            system_id,
+            unique=True,
+        ),
+    )
+
+    @classmethod
+    async def upsert_async(
+        cls: Type[Asset],
+        async_session: AsyncSession,
+        *,
+        data: Dict[str, Any],
+    ) -> Asset:
+        """
+        Creates a new Asset record if it does not exist, otherwise updates the existing Asset record
+        with the attribute values provided in the `data` dict.
+
+        Assets are looked up by the provided attributes that make up their uniqueness criteria:
+        - name
+        - asset_type
+        - domain
+        - base_url (if applicable)
+        - system_id.
+        """
+        if (
+            "name" not in data
+            or "asset_type" not in data
+            or "domain" not in data
+            or "system_id" not in data
+        ):
+            raise ValueError(
+                "name, asset_type, domain, and system_id are required fields on assets"
+            )
+
+        result = await async_session.execute(
+            select(cls).where(  # type: ignore[arg-type, call-arg]
+                cls.name == data["name"],
+                cls.asset_type == data["asset_type"],
+                cls.domain == data["domain"],
+                cls.system_id == data["system_id"],
+                cls.base_url == data.get("base_url"),
+            )
+        )
+        existing_record = result.scalars().first()
+        record_id: str
+        if existing_record:
+            await async_session.execute(
+                update(cls).where(cls.id == existing_record.id).values(data)  # type: ignore[arg-type]
```

Comment:
I haven't tried this myself just yet but do we know that this `update(...).where(...)` will properly update any field, even `meta`? Or does it have the weird behavior with sqlalchemy where changing a json property of an object and actually getting it saved requires the `flag_modified` call?

Reply:
great to call this out, i think we're OK here based on https://github.com/ethyca/fides/pull/5616/commits/9593bdd54de019275c3d02e5fed83a2b941da089